File: OrnaTrack/Stock/SuggestionsPage.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.OleDb;
using System.Drawing;
using System.Linq;
using System.Windows.Forms;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace FirstApp
{

    public partial class SuggestionsPage
    {
        public SuggestionsPage()
        {
            InitializeComponent();
            dgv = _dgv;
            _dgv.Name = "dgv";
        }
        private void SelectBT_Click(object sender, EventArgs e)
        {
            string clipboardData = Clipboard.GetText();

            string[] rows = clipboardData.Split(Conversions.ToChar(Constants.vbCrLf));

            var dt = new DataTable();
            for (int i = 0, loopTo = rows[0].Split(Conversions.ToChar(Constants.vbTab)).Length - 1; i <= loopTo; i++)
            {
                dt.Columns.Add($"Column {i}");
                var _cntr = new SuggestionControl();
                _cntr.Label1.Text = $"Column {i}:";

                _cntr.SelectBT.Click += () =>
                    {
                        var fm = new SelectAttributeForm();
                        fm.ShowDialog();
                        _cntr.CurrentAttribute = fm.SelectedAttribute;
                    };

                FlowLayoutPanel1.Controls.Add(_cntr);
            }

            for (int i = 1, loopTo1 = rows.Length - 1; i <= loopTo1; i++)
            {
                if (string.IsNullOrEmpty(rows[i].Trim()))
                    continue;

                var row = dt.NewRow();
                string[] rowCols = rows[i].Split(Conversions.ToChar(Constants.vbTab));
                for (int _idx = 0, loopTo2 = rowCols.Length - 1; _idx <= loopTo2; _idx++)
                    row[$"Column {_idx}"] = rowCols[_idx];

                dt.Rows.Add(row);
            }

            dgv.DataSource = dt;

            int totalWidth = 0;
            foreach (DataGridViewColumn col in dgv.Columns)
            {
                col.AutoSizeMode = DataGridViewAutoSizeColumnMode.AllCells;
                col.SortMode = DataGridViewColumnSortMode.NotSortable;
                totalWidth += col.Width;
            }

            Size = (Size)new Point(Size.Width + totalWidth, Size.Height);



        }

        private void SuggestionsPage_Load(object sender, EventArgs e)
        {

        }

        private void SaveBT_Click(object sender, EventArgs e)
        {

            var _mapping = new Dictionary<string, int>();

            foreach (DataGridViewColumn column in dgv.Columns)
                _mapping.Add(column.HeaderText, -1);

            int _count = 0;

            foreach (var i in FlowLayoutPanel1.Controls)
            {
                SuggestionControl _cntrl = (SuggestionControl)i;
                if (_cntrl.CurrentAttribute is null)
                {
                    MessageBox.Show("Not All Columns Are Mapped.");
                    return;
                }
                _mapping[_mapping.ElementAt(_count).Key] = _cntrl.CurrentAttribute.ID;
                _count += 1;
            }

            var schemaTable = ProjectModule.myconnection.GetOleDbSchemaTable(OleDbSchemaGuid.Columns, new object[] { null, null, "Item_Suggestion", null });

            foreach (var i in _mapping.Values)
            {
                bool _found = false;
                foreach (var _row in schemaTable.Rows)
                {
                    if ((_row.ToString() ?? "") == (i.ToString() ?? ""))
                    {
                        _found = true;
                        break;
                    }
                }
                if (_found)
                    continue;
                ProjectModule.SqlCommand($"ALTER TABLE Item_Suggestion ADD COLUMN {i} TEXT;");
            }

            for (int i = 0, loopTo = dgv.Rows.Count - 1; i <= loopTo; i++)
            {

                string _values = "";
                for (int j = 0, loopTo1 = dgv.Rows[i].Cells.Count - 1; j <= loopTo1; j++)
                {
                    var _item = dgv.Rows[i].Cells[j].Value;
                    _values = Conversions.ToString(_values + Operators.AddObject(Operators.AddObject("'", _item), "',"));
                }
                _values = _values.TrimEnd(',');


                ProjectModule.SqlCommand($"INSERT INTO Item_Suggestion ({string.Join(",", _mapping.Values)}) VALUES({_values})");
            }

            Close();

        }

    }
}

File: OrnaTrack/Stock/SuggestionsPage.vb
Imports System.Data.OleDb

Public Class SuggestionsPage
    Private Sub SelectBT_Click(sender As Object, e As EventArgs) Handles SelectBT.Click
        Dim clipboardData As String = Clipboard.GetText()

        Dim rows As String() = clipboardData.Split(vbCrLf)

        Dim dt As New DataTable
        For i = 0 To rows(0).Split(vbTab).Length - 1
            dt.Columns.Add($"Column {i}")
            Dim _cntr As New SuggestionControl
            _cntr.Label1.Text = $"Column {i}:"

            AddHandler _cntr.SelectBT.Click, Sub()
                                                 Dim fm As New SelectAttributeForm
                                                 fm.ShowDialog()
                                                 _cntr.CurrentAttribute = fm.SelectedAttribute
                                             End Sub

            FlowLayoutPanel1.Controls.Add(_cntr)
        Next

        For i As Integer = 1 To rows.Length - 1
            If rows(i).Trim() = "" Then Continue For

            Dim row As DataRow = dt.NewRow()
            Dim rowCols As String() = rows(i).Split(vbTab)
            For _idx = 0 To rowCols.Length - 1
                row($"Column {_idx}") = rowCols(_idx)
            Next

            dt.Rows.Add(row)
        Next

        dgv.DataSource = dt

        Dim totalWidth As Integer = 0
        For Each col As DataGridViewColumn In dgv.Columns
            col.AutoSizeMode = DataGridViewAutoSizeColumnMode.AllCells
            col.SortMode = DataGridViewColumnSortMode.NotSortable
            totalWidth += col.Width
        Next

        Size = New Point(Size.Width + totalWidth, Size.Height)



    End Sub

    Private Sub SuggestionsPage_Load(sender As Object, e As EventArgs) Handles MyBase.Load

    End Sub

    Private Sub SaveBT_Click(sender As Object, e As EventArgs) Handles SaveBT.Click

        Dim _mapping As New Dictionary(Of String, Integer)

        For Each column As DataGridViewColumn In dgv.Columns
            _mapping.Add(column.HeaderText, -1)
        Next

        Dim _count As Integer = 0

        For Each i In FlowLayoutPanel1.Controls
            Dim _cntrl As SuggestionControl = CType(i, SuggestionControl)
            If _cntrl.CurrentAttribute Is Nothing Then
                MessageBox.Show("Not All Columns Are Mapped.")
                Exit Sub
            End If
            _mapping(_mapping.ElementAt(_count).Key) = _cntrl.CurrentAttribute.ID
            _count += 1
        Next

        Dim schemaTable As DataTable = myconnection.GetOleDbSchemaTable(OleDbSchemaGuid.Columns,
                                                                New Object() {Nothing, Nothing, "Item_Suggestion", Nothing})

        For Each i In _mapping.Values
            Dim _found As Boolean = False
            For Each _row In schemaTable.Rows
                If _row.ToString = i.ToString Then
                    _found = True
                    Exit For
                End If
            Next
            If _found Then Continue For
            SqlCommand($"ALTER TABLE Item_Suggestion ADD COLUMN {i} TEXT;")
        Next

        For i = 0 To dgv.Rows.Count - 1

            Dim _values As String = ""
            For j = 0 To dgv.Rows(i).Cells.Count - 1
                Dim _item = dgv.Rows(i).Cells(j).Value
                _values += "'" + _item + "',"
            Next
            _values = _values.TrimEnd(","c)


            SqlCommand($"INSERT INTO Item_Suggestion ({String.Join(",", _mapping.Values)}) VALUES({_values})")
        Next

        Close()

    End Sub

End Class